lib/asset-sprayer/src/prompts.rs

```diff
@@ -0,0 +1,147 @@
+            Ok(tokio::fs::read_to_string(path).await?)
+        } else {
+            Ok(prompt.yaml_embedded().to_string())
+        }
+    }
+
+    async fn replace_prompt_request(
+        request: CreateChatCompletionRequest,
+        replace: &[(&str, &str)],
+        prompt: Prompt,
+    ) -> AssetSprayerResult<CreateChatCompletionRequest> {
+        let mut request = request;
+        for message in request.messages.iter_mut() {
+            *message =
+                Self::replace_prompt_request_message(message.clone(), replace, prompt).await?;
+        }
+        Ok(request)
+    }
+
+    async fn replace_prompt_request_message(
+        message: ChatCompletionRequestMessage,
+        replace: &[(&str, &str)],
+        prompt: Prompt,
+    ) -> AssetSprayerResult<ChatCompletionRequestMessage> {
+        let mut message = message;
+        match &mut message {
+            ChatCompletionRequestMessage::User(message) => {
+                if let ChatCompletionRequestUserMessageContent::Text(text) = &mut message.content {
+                    *text = Self::replace_prompt_text(text.clone(), replace, prompt).await?;
+                }
+            }
+            ChatCompletionRequestMessage::System(message) => {
+                if let ChatCompletionRequestSystemMessageContent::Text(text) = &mut message.content
+                {
+                    *text = Self::replace_prompt_text(text.clone(), replace, prompt).await?;
+                }
+            }
+            _ => (),
+        }
+        Ok(message)
+    }
+
+    async fn replace_prompt_text(
+        text: String,
+        replace: &[(&str, &str)],
+        prompt: Prompt,
+    ) -> AssetSprayerResult<String> {
+        let mut text = text;
+
+        // Replace {KEY} with value
+        for (from, to) in replace {
+            text = text.replace(from, to);
+        }
+
+        Self::fetch_prompt_text(&text, prompt).await
+    }
+
+    async fn fetch_prompt_text(text: &str, prompt: Prompt) -> AssetSprayerResult<String> {
+        // Fetch things between {FETCH} and {/FETCH}
+        let mut result = String::new();
+        let mut text = text;
+        while let Some(fetch_start) = text.find("{FETCH}") {
+            // Copy up to {FETCH}
+            result.push_str(&text[..fetch_start]);
+            text = &text[(fetch_start + "{FETCH}".len())..];
+
+            if let Some(url_end) = text.find("{/FETCH}") {
+                // Fetch the URL between {FETCH}...{/FETCH}
+                result.push_str(&Self::get(&text[..url_end]).await?);
+                text = &text[(url_end + "{/FETCH}".len())..];
+            } else {
+                return Err(AssetSprayerError::MissingEndFetch(prompt));
+            }
+        }
+
+        // Copy the remainder of the text
+        result.push_str(text);
+
+        Ok(result)
+    }
+
+    async fn get(url: &str) -> reqwest::Result<String> {
+        println!("Fetching: {}", url);
```

Comment:
We should formally log this

Reply:
oh lord, I'll fix this one before checkin